src/main/scala/tsiHost/TsiHostWidgetUnitTest.scala

```diff
@@ -0,0 +1,172 @@
+package testchipip
+
+import chisel3._
+import chisel3.util._
+import chisel3.experimental.dontTouch
+
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.subsystem.{BaseSubsystem}
+import freechips.rocketchip.regmapper.{HasRegMap, RegField}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.devices.tilelink.{TLTestRAM, TLROM, TLError, ErrorParams}
+import freechips.rocketchip.util._
+import freechips.rocketchip.unittest._
+
+import SerialAdapter._
+
+/**
+ * Unit test that uses the TLTSIHostWidget to interact with a target Serdesser.
+ * Currently only tests read TSI requests since TLROM can be setup easily with data
+ * that way.
+ */
+class TSIHostWidgetTest(implicit p: Parameters) extends LazyModule {
+  // params matching the configuration for the TSIHost widget
+  val beatBytes = 8
+  val lineBytes = p(PeripheryTSIHostKey).tlLineBytes
+  val serWidth = p(PeripheryTSIHostKey).serialIfWidth
+  val addrSet = p(PeripheryTSIHostKey).tlAddressSet
+  val regionTyp = p(PeripheryTSIHostKey).tlRegionType
+  val numXacts = p(PeripheryTSIHostKey).tlNumTransactions
+
+  // TSIHost widget in host-land connecting to the target Serdes
+  val hostTSIHostWidget = LazyModule(new TLTSIHostWidget(beatBytes))
+
+  // lives in target-land (connects to the TSIHost widget in host-land)
+  val targetSerdes = LazyModule(new TLSerdesser(
+    w = serWidth,
+    clientParams = TLClientParameters(
+      name = "tl-desser",
+      sourceId = IdRange(0, numXacts)),
+    managerParams = TLManagerParameters(
+      address = Seq(addrSet),
+      regionType = regionTyp,
+      supportsGet = TransferSizes(1, lineBytes),
+      supportsPutFull = TransferSizes(1, lineBytes),
+      supportsPutPartial = TransferSizes(1, lineBytes)),
+    beatBytes = beatBytes))
+
+  // rom living in target-land. connects to the client of the serdes
+  val targetRomData = Seq(
+    BigInt("7766554433221100", 16),
+    BigInt("FFEEDDCCBBAA9988", 16),
+    BigInt("0123456789ABCDEF", 16),
+    BigInt("FEDCBA9876543210", 16),
+    BigInt("76543210FDECBA98", 16))
+  val targetRom = LazyModule(new TLROM(0, 64,
+    targetRomData.flatMap(
+      data => (0 until 8).map(i => ((data >> (i * 8)) & 0xff).toByte)),
+    beatBytes = beatBytes))
+
+  // TLFragmenter splits up the multi beat packet into single beats that the left hand side can take
+  // (of size beatBytes).
+  targetRom.node := TLFragmenter(beatBytes, lineBytes) := TLBuffer() := targetSerdes.clientNode
+
+  // fuzzer that lives in target-land that sends random requests to the ram on the host
+  val targetFuzzer = LazyModule(new TLFuzzer(
```

Comment:
This is going to make it hard to write a test that will deterministically pass, because this can arbitrarily overwrite the data you are sending on TSI without your tracker knowing about it.

Reply:
In my UnitTest, shouldn't the TLFuzzer interact with the Host RAM and thus would not affect the Target RAM. So the TSI commands will not collide with the TLFuzzer transactions?